### PR TITLE
Support `maxTimeMS` argument for `find` and `findAndModify` methods

### LINKED ISSUE #497: Support `maxTimeMS` argument for `find` and `findAndModify`
`find` and `findAndModify` commands support `maxTimeMS` argument that specifies the command's time limit in milliseconds. We should support it.

The implementation should be relatively easy: get that optional parameter, and (if it is not zero) derive the Go context with a timeout and use it. In common handler code, check if the error is (possibly wrapped) context timeout, and return the right error in that case.

Of course, there are some small details that should be checked by integration tests: what types are allowed for that argument (can it be float64? int64? the document, array, null?), what error is returned on timeout, etc.

### PR DESCRIPTION
**Pull request by DoodgeMatvey**
Closes #497.

### CODE REVIEW

**Review by seeforschauer — CHANGES_REQUESTED** [review]
@DoodgeMatvey would you please add some tests or modify old tests to make it visible that now it's supported? 
Thank you.

**Review by AlekSi — CHANGES_REQUESTED** [review]
Yes, we need integration tests. For example:
* maxTimeMS=0
* maxTimeMS=-1
* maxTimeMS=1 (any very short value)
* etc

File: integration/query_test.go
Comment (DoodgeMatvey) [reply integration/query_test.go:450]: This message from mongodb

Comment (DoodgeMatvey) [reply integration/query_test.go]: Thank you, fixed

**Review by AlekSi — CHANGES_REQUESTED** [review]
There are merge conflicts

File: internal/handlers/pg/msg_find.go
Comment (AlekSi) [inline internal/handlers/pg/msg_find.go:93]: We don't need it

File: internal/handlers/pg/msg_findandmodify.go
Comment (AlekSi) [inline internal/handlers/pg/msg_findandmodify.go:214]: Why do we need that? Why can't we return errors as they are?

File: internal/handlers/pg/msg_findandmodify.go
Comment (AlekSi) [inline internal/handlers/pg/msg_findandmodify.go:363]: I think getErrorForInvalidMaxTimeMS should be modified to return the correct value too, not only error. That way, we can reduce the amount of copy&pasted code

**Review by AlekSi — CHANGES_REQUESTED** [review]

File: internal/handlers/pg/msg_findandmodify.go
Comment (seeforschauer) [inline internal/handlers/pg/msg_findandmodify.go:61]: here also.

File: internal/handlers/pg/msg_find.go
Comment (DoodgeMatvey) [reply internal/handlers/pg/msg_find.go]: fixed

File: internal/handlers/pg/msg_findandmodify.go
Comment (DoodgeMatvey) [reply internal/handlers/pg/msg_findandmodify.go]: We can, I removed it

File: internal/handlers/pg/msg_findandmodify.go
Comment (DoodgeMatvey) [reply internal/handlers/pg/msg_findandmodify.go]: Fixed

File: internal/handlers/pg/msg_findandmodify.go
Comment (DoodgeMatvey) [reply internal/handlers/pg/msg_findandmodify.go]: fixed

File: internal/handlers/pg/msg_find.go
Comment (rumyantseva) [inline internal/handlers/pg/msg_find.go:173]: Should we add some documentation like for the other unexported functions?

File: internal/handlers/pg/msg_find.go
Comment (rumyantseva) [inline internal/handlers/pg/msg_find.go:188]: "Out of range" doesn't sound specific enough to me. I'd suggest something like `maxTimeMS must not be negative (but %v is provided)`

File: internal/handlers/pg/msg_find.go
Comment (rumyantseva) [inline internal/handlers/pg/msg_find.go:195]: Same thing about documentation, let's provide something?

File: internal/handlers/pg/msg_find.go
Comment (rumyantseva) [inline internal/handlers/pg/msg_find.go:196]: Why the code above used `common.GetOptionalParam(document, "maxTimeMS", maxTimeMS)` to get the value and here it's `document.Get`? What's the difference?

File: internal/handlers/pg/msg_find.go
Comment (rumyantseva) [inline internal/handlers/pg/msg_find.go:201]: For all the error messages in this `switch`: is the difference between them so significant? Can't we just say something like `maxTimeMS must be a non-negative integer (but %v provided)` to cover all the cases in one? 

Also, the texts of those errors don't seem to be consistent. Some of them tell what maxTimeMS must be, for example, `maxTimeMS must be X`. But others are `%v value for maxTimeMS is out of range`, which doesn't give any exact details about the possible range.

To be honest, I like the  `maxTimeMS must be X` approach more because it gives the exact definition of what it should be, and the person who sees the error doesn't need to guess what's wrong with their parameter.

File: internal/handlers/pg/msg_find.go
Comment (rumyantseva) [inline internal/handlers/pg/msg_find.go:185]: Does it need any `if maxTimeMSParam > X` checks as well? Or do we accept any int32 numbers?

**Review by rumyantseva — CHANGES_REQUESTED** [review]

File: internal/handlers/pg/msg_find.go
Comment (DoodgeMatvey) [reply internal/handlers/pg/msg_find.go:196]: Because we need to check the type and return the corresponding value in the first case it is converted to type int32 and we need to check the type, so we need to get this value again and already determine the type

File: internal/handlers/pg/msg_find.go
Comment (DoodgeMatvey) [reply internal/handlers/pg/msg_find.go:185]: Yes, it does, because maxTimeMS should be non-negative integer

File: internal/handlers/pg/msg_find.go
Comment (DoodgeMatvey) [reply internal/handlers/pg/msg_find.go:188]: As i mentioned in the comment below, I tried to make the answers as much as possible the same as in mongodb, should I change this to what you suggested or can we leave the answer that would be returned when answering with mongodb?

File: internal/handlers/pg/msg_find.go
Comment (DoodgeMatvey) [reply internal/handlers/pg/msg_find.go:174]: I agree, it was added

File: internal/handlers/pg/msg_find.go
Comment (DoodgeMatvey) [reply internal/handlers/pg/msg_find.go:197]: Fixed

File: internal/handlers/pg/msg_find.go
Comment (seeforschauer) [reply internal/handlers/pg/msg_find.go:203]: hey there, 
@ribaraka @rumyantseva 
you can use `AliasFromType()`  func to make it easier.  

Here is what I did in my PR for `$set` in `internal/handlers/common/update.go` file https://github.com/FerretDB/FerretDB/pull/634/commits/e274db0512c67bdf6413e20c74243af4569e6dbc

Comment (seeforschauer) [reply internal/handlers/pg/msg_find.go:203]: `AliasFromType` was added in order to avoid unnecessary marshaling of `type.Document` in error messages. 
Here there are no documents in error messages.
However, it will decrease the complexity while the error code is the same and an error message will be still similar to mongo. 
So my opinion is that makes sense to use it here.

**Review by seeforschauer — DISMISSED** [review]
there are some comments on `AliasFromType` func in errors, which is a question for discussion on how exactly our messages should be similar to MongoDB error messages.
PR looks good.

File: internal/handlers/pg/msg_find.go
Comment (ribaraka) [reply internal/handlers/pg/msg_find.go:203]: Wouldn't it be better to rename `AliasFromType` func to `GetAliasFromType`? WDYT everyone?

Comment (seeforschauer) [reply internal/handlers/pg/msg_find.go:203]: Function names should start with a verb. `alias` is not a verb.

Comment (ribaraka) [reply internal/handlers/pg/msg_find.go:203]: Does this mean that you confirm my point of view?

Comment (seeforschauer) [reply internal/handlers/pg/msg_find.go:203]: yep 
but I suggest listening what other people say)

File: internal/handlers/pg/msg_find.go
Comment (ribaraka) [inline internal/handlers/pg/msg_find.go:201]: if you don't expect error here, you can replace with it with this:
```suggestion
	v := must.NotFail(document.Get("maxTimeMS"))
```

Comment (seeforschauer) [reply internal/handlers/pg/msg_find.go:201]: mus.NotFail will panic.
previous solution - not.

File: internal/handlers/pg/msg_find.go
Comment (DoodgeMatvey) [reply internal/handlers/pg/msg_find.go:203]: Okay, so let's discuss let's how similar our messages should be with the messages returned by mongodb. 
We can do this here or on Monday's Open Office House meeting. 
About `AliasFromType` func I prefer `GetAliasFromType` instead.

File: internal/handlers/pg/msg_find.go
Comment (DoodgeMatvey) [reply internal/handlers/pg/msg_find.go]: Fixed

File: integration/query_test.go
Comment (ribaraka) [inline integration/query_test.go:613]: Please add test case:
```
		"MaxTimeMSInt64": {
			command: bson.D{
				{"find", collection.Name()},
				{"maxTimeMS", int64(42)},
			},
		}
```

Comment (DoodgeMatvey) [reply integration/query_test.go:613]: I have `MaxTimeMSInt32` test case with `10000` value and it will be no difference with int64(42) because this value in the same range as int32. For int64 I have `BadMaxTimeMSMaxInt64` test case

File: internal/handlers/pg/msg_find.go
Comment (ribaraka) [inline internal/handlers/pg/msg_find.go:224]: how about moving the infinite cases to the ".. out of range" case?

File: internal/handlers/pg/msg_find.go
Comment (w84thesun) [inline internal/handlers/pg/msg_find.go:183]: We disard the error we get from `common.GetOptionalParam`, so this could be simplified to:
```suggestion
		return 0, getErrorForInvalidMaxTimeMS(document)
```

Comment (DoodgeMatvey) [reply internal/handlers/pg/msg_find.go]: Fixed

File: internal/handlers/pg/msg_find.go
Comment (DoodgeMatvey) [reply internal/handlers/pg/msg_find.go:218]: In such case, our messages will be very different from what mongodb returns.

File: integration/query_test.go
Comment (DoodgeMatvey) [reply integration/query_test.go:613]: Agree, added

File: integration/query_test.go
Comment (w84thesun) [inline integration/query_test.go:619]:
```suggestion
		"Double": {
			command: bson.D{
				{"find", collection.Name()},
				{"maxTimeMS", float64(10000)},
			},
		},
		"DoubleZero": {
			command: bson.D{
				{"find", collection.Name()},
				{"maxTimeMS", float64(0)},
			},
		},
		"Int32": {
			command: bson.D{
				{"find", collection.Name()},
				{"maxTimeMS", int32(10000)},
			},
		},
		"Int32Zero": {
			command: bson.D{
				{"find", collection.Name()},
				{"maxTimeMS", int32(0)},
			},
		},
		"Int64": {
			command: bson.D{
				{"find", collection.Name()},
				{"maxTimeMS", int64(10000)},
			},
		},
		"Int64Zero": {
			command: bson.D{
				{"find", collection.Name()},
				{"maxTimeMS", int64(0)},
			},
		},
```

Comment (DoodgeMatvey) [reply integration/query_test.go]: Good idea, added

**Review by w84thesun — APPROVED** [review]

File: internal/handlers/common/params.go
Comment (ribaraka) [inline internal/handlers/common/params.go:272]: What if we here (for the case of floa64) use `switch` instead of multiple `if`?

File: internal/handlers/pg/msg_find.go
Comment (DoodgeMatvey) [reply internal/handlers/pg/msg_find.go:82]: But we need to change it in this case

File: internal/handlers/common/params.go
Comment (DoodgeMatvey) [reply internal/handlers/common/params.go:297]: I think it doesn't matter here

File: internal/handlers/pg/msg_find.go
Comment (seeforschauer) [reply internal/handlers/pg/msg_find.go:82]: @DoodgeMatvey WDYM? 
it could be also applied to the code below.

Comment (DoodgeMatvey) [reply internal/handlers/pg/msg_find.go]: I see, fixed

**Review by ribaraka — APPROVED** [review]

File: internal/handlers/common/params.go
Comment (AlekSi) [inline internal/handlers/common/params.go:266]: > GetPositiveNumber returns doc's value for key or _protocol error for missing_ or invalid parameter.

Here we have a missing key but do not return a protocol error. Either description or code is wrong. I think we should return the error there. Do we have a test for that?

File: internal/handlers/common/params.go
Comment (AlekSi) [inline internal/handlers/common/params.go:263]: Let's use `GetWholeNumberParam` in that function

File: internal/handlers/pg/msg_find.go
Comment (AlekSi) [inline internal/handlers/pg/msg_find.go:78]: `run` looks out of place in this code. I think it was copied from some `run` function. It is a bad name

File: internal/handlers/pg/msg_find.go
Comment (AlekSi) [inline internal/handlers/pg/msg_find.go:81]: If `maxTimeMS` is zero we should not even create a new context that we will not use

**Review by AlekSi — CHANGES_REQUESTED** [review]

File: internal/handlers/pg/msg_findandmodify.go
Comment (seeforschauer) [inline internal/handlers/pg/msg_findandmodify.go:322]: Hey, @DoodgeMatvey  )
can you please add at least one test for `find and modify` to cover this `if` branch? Thank you.

**Comment by seeforschauer** [comment]
@DoodgeMatvey would you please resolve conflicts? Thank you.

File: internal/handlers/common/params.go
Comment (DoodgeMatvey) [reply internal/handlers/common/params.go:266]: Fixed, no we don't have a test

File: internal/handlers/common/params.go
Comment (DoodgeMatvey) [reply internal/handlers/common/params.go:263]: Fixed

File: internal/handlers/pg/msg_find.go
Comment (DoodgeMatvey) [reply internal/handlers/pg/msg_find.go]: Renamed

File: internal/handlers/pg/msg_find.go
Comment (DoodgeMatvey) [reply internal/handlers/pg/msg_find.go]: Fixed

File: internal/handlers/pg/msg_findandmodify.go
Comment (DoodgeMatvey) [reply internal/handlers/pg/msg_findandmodify.go:323]: Added

File: internal/handlers/common/params.go
Comment (DoodgeMatvey) [reply internal/handlers/common/params.go]: Fixed

File: internal/handlers/common/params.go
Comment (w84thesun) [inline internal/handlers/common/params.go:266]: Parameter `isOptional` is always true. I think it should be removed and the logic should be changed accordingly. Also it could be renamed from `GetPositiveNumber` to `GetOptionalPositiveNumber`

Comment (DoodgeMatvey) [reply internal/handlers/common/params.go]: fixed, I decided to return nil as we return in `GetOptionalParam` method

File: internal/handlers/common/params.go
Comment (w84thesun) [inline internal/handlers/common/params.go:262]:
```suggestion
// GetOptionalPositiveNumber returns doc's value for key or protocol error for invalid parameter.
```

File: integration/query_test.go
Comment (DoodgeMatvey) [reply integration/query_test.go]: Fixed

File: integration/query_test.go
Comment (DoodgeMatvey) [reply integration/query_test.go]: Fixed

File: integration/query_test.go
Comment (DoodgeMatvey) [reply integration/query_test.go]: Ok, fixed

File: integration/query_test.go
Comment (DoodgeMatvey) [reply integration/query_test.go]: Yes, fixed

File: internal/handlers/pg/msg_find.go
Comment (DoodgeMatvey) [reply internal/handlers/pg/msg_find.go]: renamed

File: internal/handlers/common/params.go
Comment (DoodgeMatvey) [reply internal/handlers/common/params.go]: fixed

File: internal/handlers/common/params.go
Comment (DoodgeMatvey) [reply internal/handlers/common/params.go:270]: fixed

**Review by w84thesun — DISMISSED** [review]

**Review by AlekSi — COMMENTED** [review]
On hold until #784 is merged. Then we should port those changes to Tigris handler.

**Comment by AlekSi** [comment]
@DoodgeMatvey The linked PR was merged. Please port those changes to the new Tigris handler.

**Review by w84thesun — APPROVED** [review]

**Review by AlekSi — APPROVED** [review]

File: internal/handlers/pg/msg_find.go
Comment (rumyantseva) [inline internal/handlers/pg/msg_find.go:82]: Yeah, this is the exact thing I expect this param to do. Why don't we have the same in `msg_findandmodify.go`?

Comment (w84thesun) [reply internal/handlers/pg/msg_find.go:82]: Now it's fixed! 😃

**Review by w84thesun — APPROVED** [review]

**Review by rumyantseva — APPROVED** [review]
LGTM

**Pull request merged**
